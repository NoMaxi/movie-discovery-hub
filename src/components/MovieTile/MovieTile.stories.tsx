import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import { MovieTile } from "./MovieTile";

const meta = {
    component: MovieTile,
    argTypes: {
        movie: {
            control: "object",
            description: "Movie data object",
        },
        onClick: {
            description: "Callback function triggered on tile click",
        },
        onEdit: {
            description: "Callback function triggered on Edit button click",
        },
        onDelete: {
            description: "Callback function triggered on Delete button click",
        },
    },
    args: {
        onClick: fn(),
        onEdit: fn(),
        onDelete: fn(),
    },
} satisfies Meta<typeof MovieTile>;

export default meta;

type Story = StoryObj<typeof MovieTile>;

export const Default: Story = {
    args: {
        movie: {
            id: "1",
            title: "Pulp Fiction",
            releaseYear: 1994,
            genres: ["All", "Horror"],
            imageUrl: "https://upload.wikimedia.org/wikipedia/en/3/3b/Pulp_Fiction_%281994%29_poster.jpg",
        },
    },
};

export const LongTitleAndGenres: Story = {
    args: {
        movie: {
            id: "2",
            title: "The Lord of the Rings: The Return of the King",
            releaseYear: 2003,
            genres: ["All", "Horror", "Comedy", "Crime", "Documentary"],
            imageUrl: "https://upload.wikimedia.org/wikipedia/commons/7/7d/El_Se%C3%B1or_de_los_Anillos_lectura.jpg",
        },
    },
};

export const NoImageAvailable: Story = {
    args: {
        movie: {
            id: "3",
            title: "Movie With Broken Image",
            releaseYear: 1994,
            genres: ["Crime", "Documentary"],
            imageUrl: "https://invalid-url-will-cause-error.abc",
        },
    },
};
