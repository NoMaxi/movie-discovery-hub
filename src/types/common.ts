export type Genre = "All" | "Documentary" | "Comedy" | "Horror" | "Crime";

export interface Movie {
    id: string;
    imageUrl: string;
    title: string;
    releaseYear: number;
    genres: string[];
}
