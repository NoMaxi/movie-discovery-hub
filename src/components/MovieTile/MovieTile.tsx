import React, { MouseEvent, useMemo, useState } from "react";
import { Movie } from "@/types/common";
import { useScrollContext } from "@/contexts/ScrollContext/useScrollContext";
import { ContextMenu } from "@/components/common/ContextMenu/ContextMenu";

interface MovieTileProps {
    movie: Movie;
    onClick: (movie: Movie, event: React.MouseEvent<HTMLDivElement>) => void;
    onEdit: (movie: Movie) => void;
    onDelete: (movie: Movie) => void;
}

export const MovieTile = ({ movie, onClick, onEdit, onDelete }: MovieTileProps) => {
    const [isContextMenuOpen, setIsContextMenuOpen] = useState(false);
    const { setTargetMovieId } = useScrollContext();

    const handleMenuToggle = (e: MouseEvent) => {
        e.stopPropagation();
        setIsContextMenuOpen(!isContextMenuOpen);
    };

    const handleTileClick = (event: React.MouseEvent<HTMLDivElement>) => {
        if (isContextMenuOpen) {
            setIsContextMenuOpen(false);
        } else {
            onClick(movie, event);
        }
    };

    const handleMovieEdit = () => {
        setTargetMovieId(movie.id);
        onEdit(movie);
    };

    const contextMenuActions = useMemo(
        () => [
            { label: "Edit", onClick: handleMovieEdit },
            { label: "Delete", onClick: () => onDelete(movie) },
        ],
        [onEdit, onDelete, movie, setTargetMovieId],
    );

    return (
        <div
            className="movie-tile relative group w-[320px] text-[var(--color-text)] cursor-pointer "
            onClick={handleTileClick}
            data-testid="movie-tile"
            data-movie-id={movie.id}
        >
            <button
                data-testid="movie-tile-menu-toggle"
                className={`
                    movie-tile-menu-toggle
                    flex items-center justify-center
                    absolute top-4 right-4 z-10 w-9 h-9
                    bg-[var(--color-gray)] bg-opacity-80 rounded-full
                    text-[var(--color-text)] text-2xl font-bold leading-none
                    opacity-0 group-hover:opacity-100 transition-opacity duration-200
                    cursor-pointer
                    ${isContextMenuOpen && "opacity-100"}
                `}
                onClick={handleMenuToggle}
            >
                &#8942;
            </button>

            <ContextMenu
                isOpen={isContextMenuOpen}
                onClose={() => setIsContextMenuOpen(false)}
                actions={contextMenuActions}
                className="w-[120px] top-4 right-14"
                withCloseButton
            />

            <img src={movie.imageUrl} alt={`${movie.title} poster`} className="w-full h-[455px] object-cover mb-4" />

            <div className="flex justify-between items-start mb-1">
                <h3 className="text-[18px] font-medium text-[var(--color-gray-lighter)] pr-2">{movie.title}</h3>
                <span
                    className="
                        flex-shrink-0
                        border border-[var(--color-gray-light)] px-3 py-1 rounded
                        text-[14px] font-medium text-[var(--color-gray-lighter)]
                    "
                >
                    {movie.releaseYear}
                </span>
            </div>
            <p className="text-[14px] text-[var(--color-gray-light)]">{movie.genres.join(", ")}</p>
        </div>
    );
};
