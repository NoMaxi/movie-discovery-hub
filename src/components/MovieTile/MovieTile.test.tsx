import { render, screen, within } from "@testing-library/react";
import userEvent, { UserEvent } from "@testing-library/user-event";
import "@testing-library/jest-dom";
import { Movie } from "@/types/common";
import MovieTile from "./MovieTile";

const mockMovie: Movie = {
    id: "1",
    title: "Test Movie Title",
    releaseYear: 1994,
    genres: ["Comedy", "Horror"],
    imageUrl: "https://upload.wikimedia.org/wikipedia/en/3/3b/Pulp_Fiction_%281994%29_poster.jpg",
};

describe("MovieTile", () => {
    let onClickMock: jest.Mock;
    let onEditMock: jest.Mock;
    let onDeleteMock: jest.Mock;
    let user: UserEvent;

    beforeEach(() => {
        onClickMock = jest.fn();
        onEditMock = jest.fn();
        onDeleteMock = jest.fn();
        user = userEvent.setup();
    });

    afterEach(() => {
        jest.clearAllMocks();
    });

    const renderComponent = () => {
        return render(
            <MovieTile movie={mockMovie} onClick={onClickMock} onEdit={onEditMock} onDelete={onDeleteMock} />,
        );
    };

    test("Should render movie tile with correct data and no context menu initially", () => {
        const { asFragment } = renderComponent();
        const img = screen.getByAltText(`${mockMovie.title} poster`);

        expect(img).toBeInTheDocument();
        expect(img).toHaveAttribute("src", mockMovie.imageUrl);

        expect(screen.getByText(mockMovie.title)).toBeInTheDocument();
        expect(screen.getByText(mockMovie.releaseYear.toString())).toBeInTheDocument();
        expect(screen.getByText(mockMovie.genres.join(", "))).toBeInTheDocument();
        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should show context menu when menu toggle button is clicked", async () => {
        renderComponent();
        const menuToggleButton = screen.getByTestId("movie-tile-menu-toggle");

        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();

        await user.click(menuToggleButton);

        const contextMenu = screen.getByTestId("context-menu");

        expect(contextMenu).toBeInTheDocument();
        expect(within(contextMenu).getByTestId("context-menu-edit-button")).toBeInTheDocument();
        expect(within(contextMenu).getByTestId("context-menu-delete-button")).toBeInTheDocument();
        expect(within(contextMenu).getByTestId("context-menu-close-button")).toBeInTheDocument();
    });

    test("Should close context menu when its close button is clicked", async () => {
        renderComponent();
        const menuToggleButton = screen.getByTestId("movie-tile-menu-toggle");

        await user.click(menuToggleButton);

        expect(screen.getByTestId("context-menu")).toBeInTheDocument();

        const closeButton = screen.getByTestId("context-menu-close-button");
        await user.click(closeButton);

        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
        expect(onEditMock).not.toHaveBeenCalled();
        expect(onDeleteMock).not.toHaveBeenCalled();
    });

    test("Should close context menu when menu toggle button is clicked while menu is open", async () => {
        renderComponent();
        const menuToggleButton = screen.getByTestId("movie-tile-menu-toggle");

        await user.click(menuToggleButton);

        expect(screen.getByTestId("context-menu")).toBeInTheDocument();

        await user.click(menuToggleButton);

        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
    });

    test("Should call 'onClick' prop with movie data when tile is clicked", async () => {
        renderComponent();
        const imageElement = screen.getByAltText(`${mockMovie.title} poster`);

        expect(imageElement).toBeInTheDocument();
        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();

        await user.click(imageElement);

        expect(onClickMock).toHaveBeenCalledTimes(1);
        expect(onClickMock).toHaveBeenCalledWith(mockMovie);
        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
    });

    test("Should call 'onEdit' prop with movie data when Edit button is clicked in context menu", async () => {
        renderComponent();
        const menuToggleButton = screen.getByTestId("movie-tile-menu-toggle");

        await user.click(menuToggleButton);

        expect(screen.getByTestId("context-menu")).toBeInTheDocument();

        const editButton = screen.getByTestId("context-menu-edit-button");
        await user.click(editButton);

        expect(onEditMock).toHaveBeenCalledTimes(1);
        expect(onEditMock).toHaveBeenCalledWith(mockMovie);
        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
    });

    test("Should call 'onDelete' prop with movie data when Delete button is clicked in context menu", async () => {
        renderComponent();
        const menuToggleButton = screen.getByTestId("movie-tile-menu-toggle");

        await user.click(menuToggleButton);
        expect(screen.getByTestId("context-menu")).toBeInTheDocument();

        const deleteButton = screen.getByTestId("context-menu-delete-button");
        await user.click(deleteButton);

        expect(screen.queryByText("Delete")).not.toBeInTheDocument();
        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
        expect(onDeleteMock).toHaveBeenCalledTimes(1);
        expect(onDeleteMock).toHaveBeenCalledWith(mockMovie);
    });

    test("Should close context menu and not call 'onClick' prop when tile is clicked while menu is open", async () => {
        renderComponent();
        const menuToggleButton = screen.getByTestId("movie-tile-menu-toggle");
        const imageElement = screen.getByAltText(`${mockMovie.title} poster`);

        await user.click(menuToggleButton);

        expect(screen.getByTestId("context-menu")).toBeInTheDocument();

        await user.click(imageElement);

        expect(screen.queryByTestId("context-menu")).not.toBeInTheDocument();
        expect(onClickMock).not.toHaveBeenCalled();
    });
});
