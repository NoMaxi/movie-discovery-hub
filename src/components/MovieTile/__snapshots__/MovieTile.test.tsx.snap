// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieTile Should render movie tile with correct data and no context menu initially 1`] = `
<DocumentFragment>
  <div
    class="movie-tile relative group w-[320px] text-[var(--color-text)] cursor-pointer "
    data-movie-id="1"
    data-testid="movie-tile"
  >
    <button
      class="
                    movie-tile-menu-toggle
                    flex items-center justify-center
                    absolute top-4 right-4 z-10 w-9 h-9
                    bg-[var(--color-gray)] bg-opacity-80 rounded-full
                    text-[var(--color-text)] text-2xl font-bold leading-none
                    opacity-0 group-hover:opacity-100 transition-opacity duration-200
                    cursor-pointer
                    false
                "
      data-testid="movie-tile-menu-toggle"
    >
      â‹®
    </button>
    <img
      alt="Pulp Fiction poster"
      class="w-full h-[455px] object-cover mb-4"
      src="https://upload.wikimedia.org/wikipedia/en/3/3b/Pulp_Fiction_%281994%29_poster.jpg"
    />
    <div
      class="flex justify-between items-start mb-1"
    >
      <h3
        class="text-[18px] font-medium text-[var(--color-gray-lighter)] pr-2"
      >
        Pulp Fiction
      </h3>
      <span
        class="
                        flex-shrink-0
                        border border-[var(--color-gray-light)] px-3 py-1 rounded
                        text-[14px] font-medium text-[var(--color-gray-lighter)]
                    "
      >
        1994
      </span>
    </div>
    <p
      class="text-[14px] text-[var(--color-gray-light)]"
    >
      Comedy, Crime
    </p>
  </div>
</DocumentFragment>
`;
